Marlin/src/module/motion.cpp

```diff
@@ -1747,10 +1747,63 @@ void homeaxis(const AxisEnum axis) {
     // so here it re-homes each tower in turn.
     // Delta homing treats the axes as normal linear axes.
 
-    // retrace by the amount specified in delta_endstop_adj + additional dist in order to have minimum steps
-    if (delta_endstop_adj[axis] * Z_HOME_DIR <= 0) {
-      if (DEBUGGING(LEVELING)) DEBUG_ECHOLNPGM("delta_endstop_adj:");
-      do_homing_move(axis, delta_endstop_adj[axis] - (MIN_STEPS_PER_SEGMENT + 1) * planner.steps_to_mm[axis] * Z_HOME_DIR);
+    #ifdef TMC_HOME_PHASE
+      static const abc_ulong_t home_phase = TMC_HOME_PHASE;
+      int axisMicrostepSize;
+      int phaseCurrent;
+      bool invertDir;
+            
+      switch (axis) {
+        case X_AXIS:
+          axisMicrostepSize = 256 / (X_MICROSTEPS);
+          phaseCurrent = stepperX.MSCNT();
+          invertDir = INVERT_X_DIR;
+          break;
+        case Y_AXIS: 
+          axisMicrostepSize = 256 / (Y_MICROSTEPS);
+          phaseCurrent = stepperY.MSCNT();
+          invertDir = INVERT_Y_DIR;
+          break;
+        case Z_AXIS: 
+          axisMicrostepSize = 256 / (Z_MICROSTEPS);
+          phaseCurrent = stepperZ.MSCNT();
+          invertDir = INVERT_Z_DIR;
+          break;
+        default: break;
+      }
+     
+      // Depending on invert dir measure the distance to nearest home phase.
+      int phaseDelta = invertDir ? phaseCurrent - home_phase[axis] : home_phase[axis] - phaseCurrent;
+
+      // Check if home distance within endstop assumed repeatability noise of .1mm and warn.
+      if ((ABS(phaseDelta) / axisMicrostepSize * planner.steps_to_mm[axis]) < 0.05f)
+        DEBUG_ECHOLNPAIR("Home phase too close to endstop trigger. Pick a different phase for ", axis_codes[axis]);
+
+      // Skip to next if target position is behind current. So it only moves away from endstop.
+      if (phaseDelta < 0) phaseDelta += 1024;
+
+      // Get the integer µsteps to target. Unreachable phase? Consistently stop at the µstep before / after based on invertDir.
+      const float mmDelta = -(int(phaseDelta / axisMicrostepSize) * planner.steps_to_mm[axis] * (Z_HOME_DIR));
+      
+      // add the delta endstop adjust
+      const float adjDistance = mmDelta + delta_endstop_adj[axis];
+
+      // optional debug messages.
+      if (DEBUGGING(LEVELING)) {
+        DEBUG_ECHOLNPAIR(
+          "Endstop ", axis_codes[axis], " hit at Phase:", phaseCurrent,
+          " Delta:", phaseDelta, " Distance:", mmDelta
+        );
+      }
+    #else
+      const float adjDistance = delta_endstop_adj[axis];
+    #endif
+
+    const float minDistance = (MIN_STEPS_PER_SEGMENT) * planner.steps_to_mm[axis];
```

Comment:
@thinkyhead Whats with this MIN_STEPS_PER_SEGMENT? Why would we not take a segment, and why the arbitrary size of 6 usteps? I see that on normal printing the next absolute move would get you to the right spot but on relative moves wont this just accumulate error?
And why is it relevant at all during homing? in here it will just add error, it is an arbitrary min threshold for endstop_adj. Can we at least remove it from homing if not  all together?

Reply:
`MIN_STEPS_PER_SEGMENT` is before my time. But the rationale is to prevent running so many tiny blocks that the machine starts to stutter from planner starvation.